wire-library/wire-protoc/src/main/java/com/squareup/wire/protocwire/WireGenerator.kt

```diff
@@ -229,42 +232,51 @@ private fun parseMessage(path: List<Int>, helper: SourceCodeHelper, packagePrefi
 private fun parseOneOfs(
   path: List<Int>, helper: SourceCodeHelper,
   oneOfDeclList: List<DescriptorProtos.OneofDescriptorProto>,
-  messageFields: MessageFields,
-  mapTypes: MutableMap<String, String>,
+  oneOfMap: Map<Int, List<FieldElement>>,
   descs: DescriptorSource
 ): List<OneOfElement> {
   val info = helper.getLocation(path)
   val result = mutableListOf<OneOfElement>()
   val oneOfPath = mutableListOf(*path.toTypedArray())
   oneOfPath.addAll(listOf(DescriptorProto.ONEOF_DECL_FIELD_NUMBER, 0))
-  val oneOfMap = mutableMapOf<Int, MutableList<Pair<Int, FieldDescriptorProto>>>()
-  for (field in messageFields.oneOfFieldList) {
-    val descriptorProto = field.second
-    if (descriptorProto.hasOneofIndex()) {
-      val oneOfIndexList = oneOfMap[descriptorProto.oneofIndex] ?: mutableListOf()
-      oneOfIndexList.add(field)
-      oneOfMap[descriptorProto.oneofIndex] = oneOfIndexList
-    }
-  }
-
   for (oneOfIndex in oneOfMap.keys) {
-    val oneOfFields = oneOfMap[oneOfIndex]!!
+    val fieldList = oneOfMap[oneOfIndex]!!
     result.add(OneOfElement(
       name = oneOfDeclList[oneOfIndex].name,
       documentation = info.comment,
-      fields = parseFields(path, helper, oneOfFields, mapTypes, descs),
+      fields = fieldList,
       groups = emptyList(),
       options = parseOptions(oneOfDeclList[oneOfIndex].options, descs)
     ))
   }
   return result
 }
 
-private fun parseFields(path: List<Int>, helper: SourceCodeHelper, fieldList: List<Pair<Int, FieldDescriptorProto>>, mapTypes: MutableMap<String, String>, descs: DescriptorSource): List<FieldElement> {
+/**
+ * The association between the FieldDescriptorProto and the FieldElement is primarily have a way to
+ * look up the oneof index of the field . The FieldElement data type loses this information on the conversion.
+ *
+ * This can be avoided if the FieldElement class contains a reference to the oneof index.
+ */
+private fun indexFieldsByOneOf(
+  fields: List<Pair<FieldDescriptorProto, FieldElement>>
+): Map<Int, List<FieldElement>> {
+  val oneOfMap = mutableMapOf<Int, MutableList<FieldElement>>()
+  for ((descriptor, fieldElement) in fields) {
+    if (descriptor.hasOneofIndex()) {
+      val list = oneOfMap[descriptor.oneofIndex] ?: mutableListOf()
+      list.add(fieldElement)
+      oneOfMap[descriptor.oneofIndex] = list
```

Comment:
Since it's mutable, you shouldn't need to put it back in the map. I think maybe above, there is a `getOrPut` thing you can use to get the value if present and put a new value if not. Then all you have to do is mutate the list to add the new item.

Reply:
OHSNAP. This method exists